gpytorch/module.py

```diff
@@ -106,68 +101,122 @@ def named_added_loss_terms(self):
+        the prior, the associated parameter names, and the transformation callable.
+
+        Yields:
+            (string, Prior, tuple((Parameter, callable)), callable): Tuple containing:
+                - the name of the prior
+                - the prior
+                - a tuple of tuples (param, transform), one for each of the parameters associated with the prior
+                - the prior's transform to be called on the parameters
+
         """
-        Adds a parameter to the module.
-        The parameter can be accessed as an attribute using given name.
+        return _extract_named_priors(module=self, memo=None, prefix="")
 
-        name (str): name of parameter
-        param (torch.nn.Parameter): parameter
-        prior (Prior): prior for parameter (default: None)
+    def named_variational_parameters(self):
+        for name, param in self.named_parameters():
+            if "variational_" in name:
+                yield name, param
+
+    def register_added_loss_term(self, name):
+        self._added_loss_terms[name] = None
+
+    def register_parameter(self, name, parameter, prior=None, transform=None):
+        r"""
+        Adds a parameter to the module. The parameter can be accessed as an attribute using the given name.
+
+        Args:
+            :attr:`name` (str):
+                The name of the parameter
+            :attr:`parameter` (torch.nn.Parameter):
+                The parameter
+            :attr:`prior` (Prior, optional):
+                The prior for parameter. Can be accessed as an attribute using <name>_prior.
+            :attr:`transform` (callable, optional):
+                The transform to be used for this parameter. Typically used for parameterizing non-negative
+                parameters such as kernel lengthscales, e.g. via softplus or exp transforms. The parameter
+                can then be transformed by calling `.transform_parameter(name, parameter)` from the module.
         """
         if "_parameters" not in self.__dict__:
             raise AttributeError("Cannot assign parameter before Module.__init__() call")
         super(Module, self).register_parameter(name, parameter)
         if prior is not None:
-            self.set_parameter_priors(**{name: prior})
+            self.register_prior("_".join([name, "prior"]), prior, name)
+        if transform is not None:
+            self.register_parameter_transform(name, transform)
 
-    def register_derived_prior(self, name, prior, parameter_names, transform):
+    def register_parameter_transform(self, parameter_name, transform):
+        if parameter_name not in self._parameters:
+            raise AttributeError(
+                "Unknown parameter {name} for {module}".format(name=parameter_name, module=self.__class__.__name__)
+            )
+        self._parameter_transforms[parameter_name] = transform
+
+    def register_prior(self, name, prior, *parameter_names, transform=None):
```

Comment:
Here - what if `register_prior` took in `name`, `prior`, and then `parameter`, which would either be a string (a name for an attribute) or a callable (some function to produce the parameters)?

For example:

- If you wanted to register a prior on `log_lengthscale` (in the log space) you could do `register_prior('lengthscale', SmoothedBoxPrior(-3, 3), "log_lengthscale")`.
- If you wanted to register a prior on `lengthscale` (not in log space), you could do `register_prior('lengthscale', SmoothedBoxPrior(0.001, 1000), "lengthscale")` - since `lengthscale` is a property on the kernel.
- You could also do `register_prior('lengthscale', SmoothedBoxPrior(0.001, 1000), lambda: return self.log_lengthscale().exp())`.

I think this would cover all the possible cases, unless I'm missing something?

Reply:
Hmm yeah that sounds like a reasonable approach. let me think through this a little more but I think I can make that change